epinar/assignment3/word_search.py

```diff
@@ -0,0 +1,128 @@
+import string
+import unittest
+
+
+class Dictionary:
+
+	def __init__(self, words, prefixes):
+		"""
+		The dictionary class to keep the values.
+		:param words: Valid entries in a dictionary.
+		:type words: Array of strings.
+		:param prefixes: Valid prefixes of each words in the dictionary.
+		:type prefixes: Array of strings.
+		"""
+		self.words = words
+		self.prefixes = prefixes
+
+	def isWord(self, s):
+		"""
+		Returns whether the given string is a valid word.
+		:param s: Word query.
+		:type s: String
+		:return: If the string is valid.
+		:rtype: bool
+		"""
+		return s in self.words
+
+	def isPrefix(self, s):
+		"""
+		Returns whether the given string is a prefix of at least one word in
+		the dictionary.
+		:param s: Prefix query.
+		:type s: String
+		:return: If the string is valid.
+		:rtype: bool
+		"""
+		return s in self.prefixes
+
+
+def word_search(grid, dict):
+	"""
+	Given a grid of letters and a dictionary, it finds all the words from the
+	dictionary that can be formed in the grid.
+	:param grid: The grid of characters to search in.
+	:type grid: 2 dimensional array of characters, M x N .
+	:param dict: The dictionary of words and prefixes.
+	:type dict: Dictionary class object.
+	:return: Set of all words found in the grid.
+	:rtype: Array of strings
+	"""
+
+	found_words = []
+	new_words = []
+	M = len(grid)
+	N = len(grid[0])
+
+	for k in range(M):
+		for l in range(N):
+			visited_cache = [[False for _ in range(N)] for _ in range(M)]
```

Comment:
visited_cache = [[False] * N] for _ in range(M)]
Yes, in Python, the list can be multiplied: the new list has the *same elements* repeated, many references to the same object instance.

See the emphasis on *same* above? If the element is immutable like False, it is ok to have many references toward it, but if the element is mutable, this won't work in ourcase, we do not want to have a list of references to the same list of False. A mutation through any reference modifies the list pointed to by all references

>>> [False] * 10
[False, False, False, False, False, False, False, False, False, False]

>>> l = [[1, 2]] * 2
[[1, 2], [1, 2]]

>>> l[0][0] = 9
[[9, 2], [9, 2]]